### ISSUE TITLE
fstar fails to build

### ISSUE THREAD

**Issue by vmrsss** [Homebrew/homebrew-core#5123]
link to output of brew gist-logs fstar

https://gist.github.com/a8e9609be396dedffa516036005b0532

The build breaks at 
opam install ocamlfind=1.5.5 batteries=2.5.2 zarith=1.3 yojson=1.1.6

it manages to download the package batteries-2.5.2, but then crashes with 

"Your request can't be satisfied:
- No package matches batteries.2.5.2.

No solution found, exiting"

**Comment by vmrsss** [comment]
anybody? thanks

**Comment by dunn** [comment]
Sorry for the late reply!  I can reproduce this, but I've tried a few different things and haven't figured out a fix yet.

**Comment by vmrsss** [comment]
thanks for the reply dunn, i am happy others can reproduce the issue, hopefully there will a fix soon. I really need FStar soon, but am very new to homebrew, i wouldn't know where to start to fix the build.

**Comment by josephlr** [comment]
It looks like there is a conflict with the dependancies of the packages being installed. `base-ocamlbuild` is needed to install `batteries`, but `base-ocamlbuild` is [will not work with `ocaml` version >= 4.03](https://opam.ocaml.org/packages/base-ocamlbuild/base-ocamlbuild.base). 

To fix this, we need to make the install sandbox use an earlier version of `ocaml`. The simple patch below fixes the build:
```diff
diff --git a/Formula/fstar.rb b/Formula/fstar.rb
index 15c471c..ee55892 100644
--- a/Formula/fstar.rb
+++ b/Formula/fstar.rb
@@ -56,6 +56,7 @@ class Fstar < Formula
       modules << "#{r.name}=#{r.version}"
     end

+    system "opam", "switch", "4.02.3"
     system "opam", "install", *modules
     system "opam", "config", "exec", "--",
     "make", "-C", "src/ocaml-output/"
```
I ran a fresh install with the patch an it worked, but someone else should test it.

However, the correct solution to this problem is to configure the sandbox to have the right version from the start, rather than switching. `opam switch` takes a very long time to complete.

Even better would be if we could fix this dependency hell, but I'm not very familiar with opam, so I'm not the right person to ask.

**Comment by ilovezfs** [comment]
Note that the fstar-on-10.12 failure has already been noted here https://github.com/Homebrew/homebrew-core/issues/5488.

**Comment by vmrsss** [comment]
thanks. isn't this an obvious issue for the package manager to fix? why are they not taking it up? is there an "official" way to ask? thank you

**Comment by ilovezfs** [comment]
@vmrsss I'll probably tweak it further, but just for you:
```
diff --git a/Formula/fstar.rb b/Formula/fstar.rb
index 15c471c..0d775c2 100644
--- a/Formula/fstar.rb
+++ b/Formula/fstar.rb
@@ -15,50 +15,24 @@ class Fstar < Formula
     sha256 "611406bfa8d13b9a6af24b066e4010b8d96dd260abec933bb215fbb180856f5e" => :mavericks
   end
 
-  depends_on "opam" => :build
+  depends_on "findutils" => :build
   depends_on "gmp" => :build
-  depends_on "ocaml" => :recommended
-  depends_on "z3" => :recommended
-
-  resource "ocamlfind" do
-    url "http://download.camlcity.org/download/findlib-1.5.5.tar.gz"
-    sha256 "aafaba4f7453c38347ff5269c6fd4f4c243ae2bceeeb5e10b9dab89329905946"
-  end
-
-  resource "batteries" do
-    url "https://github.com/ocaml-batteries-team/batteries-included/archive/v2.5.2.tar.gz"
-    sha256 "649038b47cdc2b7d4d4331fdb54b1e726212ce904c3472687a86aaa8d6006451"
-  end
-
-  resource "zarith" do
-    url "https://forge.ocamlcore.org/frs/download.php/1471/zarith-1.3.tgz"
-    sha256 "946687d6f032b96ab9db9661d876e39437bff783e0ad473ac463c06259b7a3d7"
-  end
-
-  resource "yojson" do
-    url "http://mjambon.com/releases/yojson/yojson-1.1.6.tar.gz"
-    sha256 "d97497f4d0694f515a9c12708eda2654dddb289b2ed567a377706e62e54c480d"
-  end
+  depends_on "gnu-sed" => :build
+  depends_on "ocaml" => :build
+  depends_on "opam" => :build
+  depends_on "pkg-config" => :build
+  depends_on "z3"
 
   def install
     ENV.deparallelize # Not related to F* : OCaml parallelization
 
     opamroot = buildpath/"opamroot"
     ENV["OPAMROOT"] = opamroot
-    ENV["OPAMYES"] = "1"
+    ENV["OPAMYES"] = "true"
     system "opam", "init", "--no-setup"
-    archives = opamroot/"repo/default/archives"
-    modules = []
-    resources.each do |r|
-      r.verify_download_integrity(r.fetch)
-      original_name = File.basename(r.url)
-      cp r.cached_download, archives/original_name
-      modules << "#{r.name}=#{r.version}"
-    end
-
-    system "opam", "install", *modules
-    system "opam", "config", "exec", "--",
-    "make", "-C", "src/ocaml-output/"
+    system "opam", "install", "batteries", "sqlite3", "fileutils", "stdint", "zarith", "yojson"
+    system "opam", "config", "exec", "--", "ocamlfind", "ocamlopt", "-config"
+    system "opam", "config", "exec", "--", "make", "-C", "src/ocaml-output/"
 
     bin.install "src/ocaml-output/fstar.exe"
 
```

**Comment by ilovezfs** [comment]
Fixed by #6635 in https://github.com/Homebrew/homebrew-core/commit/0e6655755b8a42cd4f8e59f5ed7e0f08980200a5

**Comment by vmrsss** [comment]
thank you!
